src/expr/src/relation/mod.rs

```diff
@@ -618,11 +619,48 @@ impl MirRelationExpr {
                 RelationType::new(base_cols).with_keys(keys)
                 // Important: do not inherit keys of either input, as not unique.
             }
-            MirRelationExpr::ArrangeBy { input, .. } => input.typ(),
-            MirRelationExpr::DeclareKeys { input, keys } => input.typ().with_keys(keys.clone()),
+            MirRelationExpr::ArrangeBy { input: _, .. } => input_types.first().unwrap().clone(),
+            MirRelationExpr::DeclareKeys { input: _, keys } => {
+                input_types.first().unwrap().clone().with_keys(keys.clone())
+            }
         }
     }
 
+    /// Reports the schema of the relation.
+    ///
+    /// This method determines the type through recursive traversal of the
+    /// relation expression, drawing from the types of base collections.
+    /// As such, this is not an especially cheap method, and should be used
+    /// judiciously.
+    pub fn typ(&self) -> RelationType {
```

Comment:
I'm not sure I follow. Breaking `typ` apart seems like a fine thing to do, but .. isn't this just as recursive as it was before, only using `type_stack` to return information? There is probably an important difference, but it is not revealed in the description or comments.

Reply:
I left the same comment that was there before, since this new method is still recursive and expensive. I'll extend the comment to explain how it works internally. Also, I'll put this one first, on top of `typ_with_input_types`.